component/main.jsonnet

```
@@ -1,10 +1,278 @@
 // The hiera parameters for the component
 local params = inv.parameters.openshift4_splunk_forwarder;
+local app_name = inv.parameters._instance;
+local app_selector = {
+  name: app_name,
+};
+
+local serviceaccount = kube.ServiceAccount(app_name);
+
+local configmap = kube.ConfigMap(app_name) {
+  data: {
+    'fluentd-loglevel': params.fluentd.loglevel,
+    'splunk-insecure': std.toString(params.splunk.insecure),
+    'splunk-hostname': params.splunk.hostname,
+    'splunk-port': std.toString(params.splunk.port),
+    'splunk-protocol': params.splunk.protocol,
+    'splunk-index': params.splunk.index,
+    'splunk-sourcetype': params.splunk.sourcetype,
+    'splunk-source': params.splunk.source,
+    'td-agent.conf': |||
+      <system>
+        log_level "#{ENV['LOG_LEVEL'] }"
+      </system>
+      <source>
+        @type  forward
+        @id    input1
+        port  24224
+        %(tls_config)s
+        <security>
+          shared_key "#{ENV['SHARED_KEY'] }"
+          self_hostname "#{ENV['HOSTNAME']}"
+        </security>
+      </source>
+      <match **>
+        @type splunk_hec
+        protocol "#{ENV['SPLUNK_PROTOCOL'] }"
+        insecure_ssl "#{ENV['SPLUNK_INSECURE'] }"
+        hec_host "#{ENV['SPLUNK_HOST'] }"
+        sourcetype "#{ENV['SPLUNK_SOURCETYPE'] }"
+        source "#{ENV['SPLUNK_SOURCE'] }"
+        index "#{ENV['SPLUNK_INDEX'] }"
+        hec_port "#{ENV['SPLUNK_PORT'] }"
+        hec_token "#{ENV['SPLUNK_TOKEN'] }"
+        host "#{ENV['NODE_NAME']}"
+        ssl_verify "#{ENV['SPLUNK_SSL_VERIFY']}"
+        ca_file /secrets/splunk/splunk-ca.crt
+        # TODO: configurize buffer config
+        <buffer>
+              @type memory
+              chunk_limit_records 100000
+              chunk_limit_size 200m
+              flush_interval 5s
+              flush_thread_count 1
+              overflow_action block
+              retry_max_times 3
+              total_limit_size 600m
+        </buffer>
+        # END: configurize buffer config
+      </match>
+    ||| % {
+      tls_config: if params.fluentd.ssl.enabled then |||
+        <transport tls>
+          cert_path /secret/fluentd/tls.crt
+          private_key_path /secret/fluentd/tls.key
+          private_key_passphrase "#{ENV['FLUENTD_SSL_PASSPHRASE'] }"
+        </transport>
+      |||,
+    },
+  },
+};
+
+local secret = kube.Secret(app_name) {
+  metadata+: {
+    labels+: {
+      'app.kubernetes.io/component': 'fluentd',
+    },
+  },
+  data: {
+    shared_key: std.base64(params.fluentd.sharedkey),
+    'hec-token': std.base64(params.splunk.token),
+    'fluentd-ssl-passsphrase': std.base64(params.fluentd.ssl.passphrase),
+    [if params.fluentd.ssl.enabled then 'forwarder-tls.crt']: std.base64(params.fluentd.ssl.cert),
+    [if params.fluentd.ssl.enabled then 'forwarder-tls.key']: std.base64(params.fluentd.ssl.key),
+    [if params.fluentd.ssl.enabled then 'ca-bundle.crt']: std.base64(params.fluentd.ssl.cert),
+  },
```

To make Vault references work, you need the following change here:

```suggestion
  stringData: {
    shared_key: params.fluentd.sharedkey,
    'hec-token': params.splunk.token,
    'fluentd-ssl-passsphrase': params.fluentd.ssl.passphrase,
    [if params.fluentd.ssl.enabled then 'forwarder-tls.crt']: params.fluentd.ssl.cert,
    [if params.fluentd.ssl.enabled then 'forwarder-tls.key']: params.fluentd.ssl.key,
    [if params.fluentd.ssl.enabled then 'ca-bundle.crt']: params.fluentd.ssl.cert,
  },
```